graph/pull.go

```diff
@@ -149,7 +149,14 @@ func (s *TagStore) pullRepository(r *registry.Session, out io.Writer, repoInfo *
 	errors := make(chan error)
 
 	layersDownloaded := false
-	for _, image := range repoData.ImgList {
+	imgIDs := []string{}
+	defer func() {
+		s.graph.Release(imgIDs, r.ID())
+	}()
+	for imgID, image := range repoData.ImgList {
+		s.graph.Retain([]string{imgID}, r.ID())
+		imgIDs = append(imgIDs, imgID)
+
 		downloadImage := func(img *registry.ImgData) {
 			if askedTag != "" && img.Tag != askedTag {
 				errors <- nil
```

Comment:
I think we should move the `Retain()` call to just after this `if` condition, because if this condition is true then we won't be downloading that image anyway, so no reason to retain it.

I'd have placed this comment a couple of lines lower but GitHub's UI won't let me it seems :(